src/mesh/cartesiangrid.jl

```diff
@@ -47,28 +47,31 @@ julia> CartesianGrid((-1.,),(1.,), dims=(100,))
 """
 struct CartesianGrid{Dim,T} <: Mesh{Dim,T}
   dims::Dims{Dim}
-  origin::Point{Dim,T}
+  reference::Point{Dim,T}
   spacing::SVector{Dim,T}
+  offset::SVector{Dim,Int}
 
-  function CartesianGrid{Dim,T}(dims, origin, spacing) where {Dim,T}
+  function CartesianGrid{Dim,T}(dims, reference, spacing, offset) where {Dim,T}
     @assert all(dims .> 0) "dimensions must be positive"
     @assert all(spacing .> 0) "spacing must be positive"
-    new(dims, origin, spacing)
+    new(dims, reference, spacing, offset)
   end
 end
 
-CartesianGrid(dims::Dims{Dim}, origin::Point{Dim,T},
-              spacing::SVector{Dim,T}) where {Dim,T} =
-  CartesianGrid{Dim,T}(dims, origin, spacing)
+CartesianGrid(dims::Dims{Dim}, reference::Point{Dim,T},
+              spacing::SVector{Dim,T},
+              offset::SVector{Dim,Int}=SVector(ntuple(i->1, Dim))) where {Dim,T} =
+  CartesianGrid{Dim,T}(dims, reference, spacing, offset)
 
-CartesianGrid(dims::Dims{Dim}, origin::NTuple{Dim,T},
-              spacing::NTuple{Dim,T}) where {Dim,T} =
-  CartesianGrid{Dim,T}(dims, Point(origin), SVector(spacing))
+CartesianGrid(dims::Dims{Dim}, reference::NTuple{Dim,T},
+              spacing::NTuple{Dim,T},
+              offset::NTuple{Dim,Int}=ntuple(i->1, Dim)) where {Dim,T} =
+  CartesianGrid{Dim,T}(dims, Point(reference), SVector(spacing), SVector(offset))
 
 function CartesianGrid(start::Point{Dim,T}, finish::Point{Dim,T},
                        spacing::SVector{Dim,T}) where {Dim,T}
   dims = Tuple(ceil.(Int, (finish - start) ./ spacing))
-  CartesianGrid{Dim,T}(dims, start, spacing)
+  CartesianGrid{Dim,T}(dims, start, spacing, SVector(ntuple(i->1, Dim)))
```

Comment:
Please define the offset variable outside the constructor for improved readability.